### PR TITLE
Add z poly z

### PR DESCRIPTION
**Pull request by 3pmile**
**Description**

Addition for Z and PolyOverZ

This PR implements...
- [x]  Z + PolyOverZ (and all burrowed versions)
- [ ]  PolyOverZ + Z (and all burrowed versions)

for/ of `Component`.

Addition

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [ ] I have performed a self-review of my own code
  - [ ] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [ ] The chosen implementation is not more complex than it has to be
  - [ ] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [ ] The doc comments fit our style guide
  - [ ] I have credited related sources if needed

### CODE REVIEW

**Comment by 3pmile** [comment]
Please have a look at the implementation and whether edge cases for tests are missing

File: src/integer/z/arithmetic/add.rs [hunk 1]
```diff
@@ -140,6 +144,51 @@ impl Add<&Zq> for &Z {
...
+            length: 0,
+        };
+        unsafe {
+            fmpz_poly_set(&mut out, &other.poly);
+            fmpz_poly_get_coeff_fmpz(&mut added, &other.poly, 0);
+            fmpz_add(&mut added, &self.value, &added);
+            fmpz_poly_set_coeff_fmpz(&mut out, 0, &added);
+        }
+        PolyOverZ { poly: out }
+    }
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:186]:
```suggestion
        // check if the first coefficient of the polynomial is initiated and
        // can be addressed
        if other.is_zero() {
            PolyOverZ::from(self)
        } else {
            let out = other.clone();
            unsafe {
                fmpz_add(out.poly.coeffs, &self.value, out.poly.coeffs);
            }
            out
        }
```

I think this solution is cleaner as we only have to interact with FLINT itslef minimally. Also, your `added` is never cleared which would cause a memory-leak

Comment (Marvin-Beckmann) [reply src/integer/z/arithmetic/add.rs:186]: poly.coeffs adresses the first fmpz in the list of coefficients, therefore we can address the first coefficient this way (if defined)

File: src/integer/z/arithmetic/add.rs [hunk 2]
```diff
@@ -21,6 +22,9 @@ use flint_sys::{
     fmpq::fmpq_add_fmpz,
     fmpz::{fmpz, fmpz_add},
     fmpz_mod::fmpz_mod_add_fmpz,
+    fmpz_poly::{
+        fmpz_poly_get_coeff_fmpz, fmpz_poly_set, fmpz_poly_set_coeff_fmpz, fmpz_poly_struct,
+    },
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:27]: probably dont need all when adapting the changes

File: src/integer/z/arithmetic/add.rs [hunk 3]
```diff
@@ -398,3 +447,68 @@ mod test_add_between_z_and_zq {
...
+    }
+
+    /// Testing addition for [`Z`] and borrowed [`PolyOverZ`]
+    #[test]
+    fn add_second_borrowed() {
+        let a: Z = Z::from(9);
+        let b: PolyOverZ = PolyOverZ::from_str("4  1 1 2 3").unwrap();
+        let c: PolyOverZ = a + &b;
+        assert_eq!(c, PolyOverZ::from_str("4  10 1 2 3").unwrap());
+    }
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:491]: possibly batch them for brevity and call the test itself availability as we have done for some other types

@Marcel583 can probably look for what is most consistent, but I would opt for a less repetititve design (at least right now I would)

File: src/integer/z/arithmetic/add.rs [hunk 4]
```diff
@@ -398,3 +447,68 @@ mod test_add_between_z_and_zq {
...
+
+    /// Testing addition for an empty polynomial
+    #[test]
+    fn add_zero_polynomial() {
+        let a: Z = Z::from(15);
+        let b: PolyOverZ = PolyOverZ::default();
+        let c: PolyOverZ = a + b;
+        assert_eq!(c, PolyOverZ::from_str("1  15").unwrap());
+    }
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:513]: I would also add the edgecase where you add a Z::ZERO (for consistency in both directions)

File: src/integer/z/arithmetic/add.rs [hunk 5]
```diff
@@ -140,6 +144,51 @@ impl Add<&Zq> for &Z {
...
+    /// use std::str::FromStr;
+    ///
+    /// let a: Z = Z::from(42);
+    /// let b: PolyOverZ = PolyOverZ::from_str("4  5 1 2 3").unwrap();
+    ///
+    /// let c: PolyOverZ = &a + &b;
+    /// let d: PolyOverZ = a + b;
+    /// let e: PolyOverZ = &Z::from(42) + d;
+    /// let f: PolyOverZ = Z::from(42) + &e;
+    /// ```
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:170]: maybe add an `assert_eq!(...)` where the reader can see the actual addition
```suggestion
    /// ```
    /// use qfall_math::integer::PolyOverZ;
    /// use qfall_math::integer::Z;
    /// use std::str::FromStr;
    ///
    /// let a: Z = Z::from(42);
    /// let b: PolyOverZ = PolyOverZ::from_str("4  5 1 2 3").unwrap();
    ///
    /// let c: PolyOverZ = &a + &b;
    /// let d: PolyOverZ = a + b;
    /// let e: PolyOverZ = &Z::from(42) + d;
    /// let f: PolyOverZ = Z::from(42) + &e;
    /// 
    /// assert_eq!(PolyOverZ::from_str!("4  47 1 2 3").unwrap(), a+b)
    /// ```
```
Maybe choose some clever way so the expected behaviour is shown and not just the available types

File: src/integer/z/arithmetic/add.rs (hunk 1 above)
Comment (3pmile) [reply src/integer/z/arithmetic/add.rs:186]: applied

File: src/integer/z/arithmetic/add.rs (hunk 2 above)
Comment (3pmile) [reply src/integer/z/arithmetic/add.rs:27]: cleared

File: src/integer/z/arithmetic/add.rs (hunk 4 above)
Comment (3pmile) [reply src/integer/z/arithmetic/add.rs:513]: added

File: src/integer/z/arithmetic/add.rs (hunk 3 above)
Comment (3pmile) [reply src/integer/z/arithmetic/add.rs:491]: I have talked to @Marcel583 who said this is the consistent solution for now, but we should have a look at all test in a refactoring

File: src/integer/z/arithmetic/add.rs (hunk 5 above)
Comment (3pmile) [reply src/integer/z/arithmetic/add.rs:170]: in alinement with the other tests I will keep the current implementation

**Review by Marvin-Beckmann — APPROVED** [review]
